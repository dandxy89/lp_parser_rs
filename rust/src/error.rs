//! Comprehensive error types for LP parsing operations.
//!
//! This module provides a structured error hierarchy with detailed context
//! information for debugging and error handling in LP file parsing.

use thiserror::Error;

/// Comprehensive error type for LP parsing operations.
///
/// This error type provides detailed context about parsing failures,
/// including location information and specific error conditions.
#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum LpParseError {
    /// Invalid or malformed constraint syntax
    #[error("Invalid constraint syntax at position {position}: {context}")]
    ConstraintSyntax { position: usize, context: String },

    /// Invalid or malformed objective syntax
    #[error("Invalid objective syntax at position {position}: {context}")]
    ObjectiveSyntax { position: usize, context: String },

    /// Unknown or invalid variable type specification
    #[error("Unknown variable type '{var_type}' for variable '{variable}'")]
    UnknownVariableType { variable: String, var_type: String },

    /// Reference to an undefined variable
    #[error("Undefined variable '{variable}' referenced in {context}")]
    UndefinedVariable { variable: String, context: String },

    /// Duplicate definition of a component
    #[error("Duplicate {component_type} '{name}' defined")]
    DuplicateDefinition { component_type: String, name: String },

    /// Invalid numerical value or format
    #[error("Invalid number format '{value}' at position {position}")]
    InvalidNumber { value: String, position: usize },

    /// Missing required section in LP file
    #[error("Missing required section: {section}")]
    MissingSection { section: String },

    /// Invalid bound specification
    #[error("Invalid bounds for variable '{variable}': {details}")]
    InvalidBounds { variable: String, details: String },

    /// Invalid SOS constraint specification
    #[error("Invalid SOS constraint '{constraint}': {details}")]
    InvalidSosConstraint { constraint: String, details: String },

    /// Validation error for logical consistency
    #[error("Validation error: {message}")]
    ValidationError { message: String },

    /// Generic parsing error with context
    #[error("Parse error at position {position}: {message}")]
    ParseError { position: usize, message: String },

    /// File I/O related errors
    #[error("File I/O error: {message}")]
    IoError { message: String },

    /// Internal parser state errors
    #[error("Internal parser error: {message}")]
    InternalError { message: String },
}

impl LpParseError {
    /// Create a new constraint syntax error
    pub fn constraint_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ConstraintSyntax { position, context: context.into() }
    }

    /// Create a new objective syntax error
    pub fn objective_syntax(position: usize, context: impl Into<String>) -> Self {
        Self::ObjectiveSyntax { position, context: context.into() }
    }

    /// Create a new unknown variable type error
    pub fn unknown_variable_type(variable: impl Into<String>, var_type: impl Into<String>) -> Self {
        Self::UnknownVariableType { variable: variable.into(), var_type: var_type.into() }
    }

    /// Create a new undefined variable error
    pub fn undefined_variable(variable: impl Into<String>, context: impl Into<String>) -> Self {
        Self::UndefinedVariable { variable: variable.into(), context: context.into() }
    }

    /// Create a new duplicate definition error
    pub fn duplicate_definition(component_type: impl Into<String>, name: impl Into<String>) -> Self {
        Self::DuplicateDefinition { component_type: component_type.into(), name: name.into() }
    }

    /// Create a new invalid number error
    pub fn invalid_number(value: impl Into<String>, position: usize) -> Self {
        Self::InvalidNumber { value: value.into(), position }
    }

    /// Create a new missing section error
    pub fn missing_section(section: impl Into<String>) -> Self {
        Self::MissingSection { section: section.into() }
    }

    /// Create a new invalid bounds error
    pub fn invalid_bounds(variable: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidBounds { variable: variable.into(), details: details.into() }
    }

    /// Create a new invalid SOS constraint error
    pub fn invalid_sos_constraint(constraint: impl Into<String>, details: impl Into<String>) -> Self {
        Self::InvalidSosConstraint { constraint: constraint.into(), details: details.into() }
    }

    /// Create a new validation error
    pub fn validation_error(message: impl Into<String>) -> Self {
        Self::ValidationError { message: message.into() }
    }

    /// Create a new parse error
    pub fn parse_error(position: usize, message: impl Into<String>) -> Self {
        Self::ParseError { position, message: message.into() }
    }

    /// Create a new I/O error
    pub fn io_error(message: impl Into<String>) -> Self {
        Self::IoError { message: message.into() }
    }

    /// Create a new internal error
    pub fn internal_error(message: impl Into<String>) -> Self {
        Self::InternalError { message: message.into() }
    }
}

/// Convert from nom parsing errors to our custom error type
impl<'a> From<nom::Err<nom::error::Error<&'a str>>> for LpParseError {
    fn from(err: nom::Err<nom::error::Error<&'a str>>) -> Self {
        match err {
            nom::Err::Incomplete(_) => Self::parse_error(0, "Incomplete input"),
            nom::Err::Error(e) | nom::Err::Failure(e) => Self::parse_error(e.input.as_ptr() as usize, format!("Parse error: {:?}", e.code)),
        }
    }
}

/// Convert from standard I/O errors
impl From<std::io::Error> for LpParseError {
    fn from(err: std::io::Error) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Convert from boxed errors (used by CSV module)
impl From<Box<dyn std::error::Error>> for LpParseError {
    fn from(err: Box<dyn std::error::Error>) -> Self {
        Self::io_error(err.to_string())
    }
}

/// Result type alias for LP parsing operations
pub type LpResult<T> = Result<T, LpParseError>;

/// Context extension trait for adding location information to errors
pub trait ErrorContext<T> {
    /// Add position context to an error
    fn with_position(self, position: usize) -> LpResult<T>;

    /// Add general context to an error
    fn with_context(self, context: &str) -> LpResult<T>;
}

impl<T> ErrorContext<T> for Result<T, LpParseError> {
    fn with_position(self, position: usize) -> Self {
        self.map_err(|mut err| {
            if let LpParseError::ParseError { position: ref mut pos, .. } = &mut err {
                *pos = position;
            }
            err
        })
    }

    fn with_context(self, context: &str) -> Self {
        self.map_err(|err| match err {
            LpParseError::ParseError { position, message } => LpParseError::parse_error(position, format!("{context}: {message}")),
            other => other,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_creation() {
        let err = LpParseError::constraint_syntax(42, "missing operator");
        assert_eq!(err.to_string(), "Invalid constraint syntax at position 42: missing operator");
    }

    #[test]
    fn test_error_context() {
        let result: LpResult<()> = Err(LpParseError::parse_error(10, "test error"));
        let with_context = result.with_context("parsing constraint");

        assert!(with_context.is_err());
        assert!(with_context.unwrap_err().to_string().contains("parsing constraint"));
    }

    #[test]
    fn test_io_error_conversion() {
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let lp_err: LpParseError = io_err.into();

        match lp_err {
            LpParseError::IoError { message } => assert!(message.contains("file not found")),
            _ => panic!("Expected IoError"),
        }
    }
}
