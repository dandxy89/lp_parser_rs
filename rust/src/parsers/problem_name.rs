//! Parser for LP problem names and comments.
//!
//! This module handles parsing of problem names that typically
//! appear in comments at the start of LP files.
//!

use std::borrow::Cow;

use nom::branch::alt;
use nom::bytes::complete::{tag, take_until};
use nom::character::complete::{line_ending, multispace0, not_line_ending};
use nom::combinator::recognize;
use nom::multi::many0;
use nom::{IResult, Parser as _};

#[inline]
/// Parses a single comment from the input string.
///
/// This function recognises three types of comment starts: `\\*`, `\*`, and `\`.
/// For block comments starting with `\\*` or `\*`, it captures content until `*\` is found.
/// For line comments starting with `\`, it captures content until the end of the line.
///
/// # Arguments
///
/// * `input` - A string slice that holds the input to be parsed.
///
/// # Returns
///
/// An `IResult` containing the remaining input and the parsed comment content.
///
fn parse_single_comment(input: &str) -> IResult<&str, &str> {
    let (input, comment_start) = alt((tag("\\\\*"), tag("\\*"), tag("\\"))).parse(input)?;
    let (input, content) = match comment_start {
        "\\\\*" | "\\*" => {
            let (i, content) = recognize(take_until("*\\")).parse(input)?;
            let (i, _) = (tag("*\\"), multispace0).parse(i)?;
            (i, content)
        }
        "\\" => {
            let (i, content) = recognize(not_line_ending).parse(input)?;
            let (i, _) = line_ending(i)?;
            (i, content)
        }
        _ => unreachable!(),
    };
    Ok((input, content))
}

#[inline]
/// Extracts the last comment (if present) from a sequence of comments in the input string.
///
/// # Arguments
///
/// * `input` - A string slice that holds the input data to be parsed.
///
/// # Returns
///
/// * `IResult<&str, Option<&str>>` - A result containing the remaining unparsed input and an
///   optional string slice of the last comment found. If no comments are found, returns `None`.
///
pub fn parse_problem_name(input: &str) -> IResult<&str, Option<Cow<'_, str>>> {
    let (remaining, comments) = many0(parse_single_comment).parse(input)?;
    let problem_name = comments.last().copied().map(Cow::Borrowed);
    Ok((remaining, problem_name))
}

#[cfg(test)]
mod test {
    use crate::parsers::problem_name::parse_problem_name;

    #[test]
    fn test_parse_lp_file_comments() {
        let valid = [
            "\\* First comment *\\\n\\ENCODING=ISO-8859-1\n\\* Middle comment *\\\\Problem name: ilog.cplex\n\\* Last comment *\\",
            "\\Problem name: kb2.mps\n",
            "\\ File: lo1.lp\n",
            "\\* WBM_Problem *\\\n",
        ];
        for input in valid {
            let (remainder, p_name) = parse_problem_name(input).unwrap();
            assert_eq!("", remainder);
            assert!(p_name.is_some());
        }
    }
}
