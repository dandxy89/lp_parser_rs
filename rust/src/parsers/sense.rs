//! Parser for optimization sense in LP files.
//!

use nom::branch::alt;
use nom::bytes::complete::tag_no_case;
use nom::character::complete::multispace0;
use nom::combinator::value;
use nom::sequence::delimited;
use nom::{IResult, Parser as _};

use crate::model::Sense;

#[inline]
/// Parses the input string to determine the optimization sense.
///
/// This function attempts to match the input string against known
/// optimization sense keywords, such as "minimize", "maximum", etc.,
/// and returns the corresponding `Sense` variant. It ignores leading
/// and trailing whitespace and is case-insensitive.
///
/// # Arguments
///
/// * `input` - A string slice that holds the input to be parsed.
///
/// # Returns
///
/// * `IResult<&str, Sense>` - A result containing the remaining input
///   and the parsed `Sense` variant if successful, or an error if parsing fails.
///
pub fn parse_sense(input: &str) -> IResult<&str, Sense> {
    delimited(
        multispace0,
        alt((
            value(Sense::Minimize, alt((tag_no_case("minimize"), tag_no_case("minimise"), tag_no_case("minimum"), tag_no_case("min")))),
            value(Sense::Maximize, alt((tag_no_case("maximize"), tag_no_case("maximise"), tag_no_case("maximum"), tag_no_case("max")))),
        )),
        multispace0,
    )
    .parse(input)
}

#[cfg(test)]
mod test {
    use crate::parsers::sense::parse_sense;

    #[test]
    fn test_parse_sense() {
        let valid = ["Minimize", "minimize", "minimise", "min", "minimum", "Maximize", "maximize", "maximise", "Max", "maximum"];
        for input in valid {
            assert!(parse_sense(input).is_ok());
        }
    }
}
