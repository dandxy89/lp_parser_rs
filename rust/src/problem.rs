//! Main problem representation and parsing logic.
//!
//! This module defines the `LpProblem` struct and it's associated
//! parsing functionality. It serves as the main entry point for
//! working with LP problems.
//!

use std::borrow::Cow;
use std::collections::HashMap;
use std::collections::hash_map::Entry;

use nom::Parser as _;
use nom::combinator::opt;

use crate::error::{LpParseError, LpResult};
use crate::model::{Constraint, Objective, Sense, Variable, VariableType};
use crate::parsers::constraint::{parse_constraint_header, parse_constraints};
use crate::parsers::objective::parse_objectives;
use crate::parsers::problem_name::parse_problem_name;
use crate::parsers::sense::parse_sense;
use crate::parsers::sos_constraint::parse_sos_section;
use crate::parsers::variable::{
    parse_binary_section, parse_bounds_section, parse_generals_section, parse_integer_section, parse_semi_section,
};
use crate::{
    ALL_BOUND_HEADERS, BINARY_HEADERS, CONSTRAINT_HEADERS, END_HEADER, GENERAL_HEADERS, INTEGER_HEADERS, SEMI_HEADERS, SOS_HEADERS,
    is_binary_section, is_bounds_section, is_generals_section, is_integers_section, is_semi_section, is_sos_section, take_until_parser,
};

// Type aliases to reduce complexity
type ObjectivesParseResult<'a> = (HashMap<Cow<'a, str>, Objective<'a>>, HashMap<&'a str, Variable<'a>>);
type ConstraintsParseResult<'a> = (&'a str, HashMap<Cow<'a, str>, Constraint<'a>>, HashMap<&'a str, Variable<'a>>);

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[derive(Debug, Default, PartialEq)]
/// Represents a Linear Programming (LP) problem.
///
/// The `LpProblem` struct encapsulates the components of an LP problem, including its name,
/// sense (e.g., minimization, or maximization), objectives, constraints, and variables.
///
/// # Attributes
///
/// * `#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]`:
///   Enables the `diff` feature for comparing differences between instances of `LpProblem`.
/// * `#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]`:
///   Enables serialization and deserialization of `LpProblem` instances when the `serde` feature is active.
///
pub struct LpProblem<'a> {
    /// An optional reference to a string slice representing the name of the LP problem.
    pub name: Option<Cow<'a, str>>,
    /// The optimization sense of the problem, indicating whether it is a minimization or maximization problem.
    pub sense: Sense,
    /// A `HashMap` where the keys are the names of the objectives and the values are `Objective` structs.
    pub objectives: HashMap<Cow<'a, str>, Objective<'a>>,
    /// A `HashMap` where the keys are the names of the constraints and the values are `Constraint` structs.
    pub constraints: HashMap<Cow<'a, str>, Constraint<'a>>,
    /// A `HashMap` where the keys are the names of the variables and the values are `Variable` structs.
    pub variables: HashMap<&'a str, Variable<'a>>,
}

impl<'a> LpProblem<'a> {
    #[must_use]
    #[inline]
    /// Initialise a new `Self`
    pub fn new() -> Self {
        Self::default()
    }

    #[must_use]
    #[inline]
    /// Override the problem name
    pub fn with_problem_name(self, problem_name: Cow<'a, str>) -> Self {
        Self { name: Some(problem_name), ..self }
    }

    #[must_use]
    #[inline]
    /// Override the problem sense
    pub fn with_sense(self, sense: Sense) -> Self {
        Self { sense, ..self }
    }

    #[must_use]
    #[inline]
    /// Returns the name of the LP Problem
    pub fn name(&self) -> Option<&str> {
        self.name.as_deref()
    }

    #[must_use]
    #[inline]
    /// Returns `true` if the `Self` a Minimize LP Problem
    pub fn is_minimization(&self) -> bool {
        self.sense.is_minimization()
    }

    #[must_use]
    #[inline]
    /// Returns the number of constraints contained within the Problem
    pub fn constraint_count(&self) -> usize {
        self.constraints.len()
    }

    #[must_use]
    #[inline]
    /// Returns the number of objectives contained within the Problem
    pub fn objective_count(&self) -> usize {
        self.objectives.len()
    }

    #[must_use]
    #[inline]
    /// Returns the number of variables contained within the Problem
    pub fn variable_count(&self) -> usize {
        self.variables.len()
    }

    #[inline]
    /// Parse a `Self` from a string slice
    pub fn parse(input: &'a str) -> LpResult<Self> {
        log::debug!("Starting to parse LP problem");
        Self::try_from(input)
    }

    #[inline]
    /// Add a new variable to the problem.
    ///
    /// If a variable with the same name already exists, it will be replaced.
    pub fn add_variable(&mut self, variable: Variable<'a>) {
        self.variables.insert(variable.name, variable);
    }

    #[inline]
    /// Add a new constraint to the problem.
    ///
    /// If a constraint with the same name already exists, it will be replaced.
    pub fn add_constraint(&mut self, constraint: Constraint<'a>) {
        let name = constraint.name().as_ref().to_owned();

        if let Constraint::Standard { coefficients, .. } = &constraint {
            for coeff in coefficients {
                if !self.variables.contains_key(coeff.name) {
                    self.variables.insert(coeff.name, Variable::new(coeff.name));
                }
            }
        }

        if let Constraint::SOS { weights, .. } = &constraint {
            for coeff in weights {
                if !self.variables.contains_key(coeff.name) {
                    self.variables.insert(coeff.name, Variable::new(coeff.name).with_var_type(VariableType::SOS));
                }
            }
        }

        self.constraints.insert(Cow::Owned(name), constraint);
    }

    #[inline]
    /// Add a new objective to the problem.
    ///
    /// If an objective with the same name already exists, it will be replaced.
    pub fn add_objective(&mut self, objective: Objective<'a>) {
        for coeff in &objective.coefficients {
            if !self.variables.contains_key(coeff.name) {
                self.variables.insert(coeff.name, Variable::new(coeff.name));
            }
        }

        let name = objective.name.clone();
        self.objectives.insert(name, objective);
    }
}

impl std::fmt::Display for LpProblem<'_> {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(problem_name) = &self.name {
            writeln!(f, "Problem name: {problem_name}")?;
        }
        writeln!(f, "Sense: {}", self.sense)?;
        writeln!(f, "Objectives: {}", self.objectives.len())?;
        writeln!(f, "Constraints: {}", self.constraints.len())?;
        writeln!(f, "Variables: {}", self.variables.len())?;

        Ok(())
    }
}

impl<'a> LpProblem<'a> {
    /// Parse the header section (name and sense)
    fn parse_header_section(input: &'a str) -> LpResult<(&'a str, &'a str, Option<Cow<'a, str>>, Sense)> {
        let (remaining_input, (name, sense, obj_section, ())) =
            (parse_problem_name, parse_sense, take_until_parser(&CONSTRAINT_HEADERS), parse_constraint_header)
                .parse(input)
                .map_err(|err| LpParseError::parse_error(0, format!("Failed to parse header section: {err:?}")))?;

        Ok((remaining_input, obj_section, name, sense))
    }

    /// Parse the objectives section
    fn parse_objectives_section(input: &'a str) -> LpResult<ObjectivesParseResult<'a>> {
        let (_, (objectives, variables)) =
            parse_objectives(input).map_err(|err| LpParseError::objective_syntax(0, format!("Failed to parse objectives: {err:?}")))?;

        Ok((objectives, variables))
    }

    /// Parse the constraints section
    fn parse_constraints_section(input: &'a str) -> LpResult<ConstraintsParseResult<'a>> {
        let (input, constraint_str) = take_until_parser(&ALL_BOUND_HEADERS)(input)
            .map_err(|err| LpParseError::constraint_syntax(0, format!("Failed to find constraints section: {err:?}")))?;

        let (_, (constraints, constraint_vars)) = parse_constraints(constraint_str)
            .map_err(|err| LpParseError::constraint_syntax(0, format!("Failed to parse constraints: {err:?}")))?;

        Ok((input, constraints, constraint_vars))
    }

    /// Parse variable bounds section
    fn parse_bounds_section(input: &'a str, variables: &mut HashMap<&'a str, Variable<'a>>) -> LpResult<&'a str> {
        if is_bounds_section(input).is_ok() {
            let (rem_input, bound_str) = take_until_parser(&INTEGER_HEADERS)(input)
                .map_err(|err| LpParseError::parse_error(0, format!("Failed to parse bounds section: {err:?}")))?;

            let (_, bounds) = parse_bounds_section(bound_str)
                .map_err(|err| LpParseError::invalid_bounds("unknown", format!("Failed to parse bounds: {err:?}")))?;

            for (name, var_type) in bounds {
                match variables.entry(name) {
                    Entry::Occupied(mut occupied_entry) => {
                        occupied_entry.get_mut().set_var_type(var_type);
                    }
                    Entry::Vacant(vacant_entry) => {
                        vacant_entry.insert(Variable { name, var_type });
                    }
                }
            }

            Ok(rem_input)
        } else {
            Ok(input)
        }
    }

    /// Parse variable type sections (integers, generals, binaries, semi-continuous)
    fn parse_variable_type_sections(mut input: &'a str, variables: &mut HashMap<&'a str, Variable<'a>>) -> LpResult<&'a str> {
        // Integer
        if is_integers_section(input).is_ok() {
            if let Ok((rem_input, Some(integer_str))) = opt(take_until_parser(&GENERAL_HEADERS)).parse(input) {
                if let Ok((_, integer_vars)) = parse_integer_section(integer_str) {
                    set_var_types(variables, integer_vars, VariableType::Integer);
                }
                input = rem_input;
            }
        }

        // General
        if is_generals_section(input).is_ok() {
            if let Ok((rem_input, Some(generals_str))) = opt(take_until_parser(&BINARY_HEADERS)).parse(input) {
                if let Ok((_, general_vars)) = parse_generals_section(generals_str) {
                    set_var_types(variables, general_vars, VariableType::General);
                }
                input = rem_input;
            }
        }

        // Binary
        if is_binary_section(input).is_ok() {
            if let Ok((rem_input, Some(binary_str))) = opt(take_until_parser(&SEMI_HEADERS)).parse(input) {
                if let Ok((_, binary_vars)) = parse_binary_section(binary_str) {
                    set_var_types(variables, binary_vars, VariableType::Binary);
                }
                input = rem_input;
            }
        }

        // Semi-continuous
        if is_semi_section(input).is_ok() {
            if let Ok((rem_input, Some(semi_str))) = opt(take_until_parser(&SOS_HEADERS)).parse(input) {
                if let Ok((_, semi_vars)) = parse_semi_section(semi_str) {
                    set_var_types(variables, semi_vars, VariableType::SemiContinuous);
                }
                input = rem_input;
            }
        }

        Ok(input)
    }

    /// Parse SOS constraints section
    fn parse_sos_section(
        input: &'a str,
        constraints: &mut HashMap<Cow<'a, str>, Constraint<'a>>,
        variables: &mut HashMap<&'a str, Variable<'a>>,
    ) -> LpResult<&'a str> {
        if is_sos_section(input).is_ok() {
            if let Ok((rem_input, Some(sos_str))) = opt(take_until_parser(&END_HEADER)).parse(input) {
                if let Ok((_, Some((sos_constraints, constraint_vars)))) = opt(parse_sos_section).parse(sos_str) {
                    variables.extend(constraint_vars);
                    for (name, constraint) in sos_constraints {
                        constraints.insert(name, constraint);
                    }
                }
                Ok(rem_input)
            } else {
                Ok(input)
            }
        } else {
            Ok(input)
        }
    }

    /// Validate remaining unparsed input
    fn validate_remaining_input(input: &str) -> LpResult<()> {
        if input.len() > 3 {
            log::warn!("Unused input not parsed by `LpProblem`: {input}");
        }
        Ok(())
    }
}

#[inline]
fn set_var_types<'a>(variables: &mut HashMap<&'a str, Variable<'a>>, vars: Vec<&'a str>, var_type: VariableType) {
    for name in vars {
        match variables.entry(name) {
            Entry::Occupied(mut occupied_entry) => {
                occupied_entry.get_mut().set_var_type(var_type.clone());
            }
            Entry::Vacant(vacant_entry) => {
                vacant_entry.insert(Variable { name, var_type: var_type.clone() });
            }
        }
    }
}

impl<'a> TryFrom<&'a str> for LpProblem<'a> {
    type Error = LpParseError;

    #[inline]
    fn try_from(input: &'a str) -> Result<Self, Self::Error> {
        log::debug!("Starting to parse LP problem");

        // Parse header section (name and sense)
        let (input, obj_section, name, sense) = Self::parse_header_section(input)?;

        // Parse objectives section
        let (objectives, mut variables) = Self::parse_objectives_section(obj_section)?;

        // Parse constraints section
        let (mut input, mut constraints, constraint_vars) = Self::parse_constraints_section(input)?;
        variables.extend(constraint_vars);

        // Parse bounds section
        input = Self::parse_bounds_section(input, &mut variables)?;

        // Parse variable type sections
        input = Self::parse_variable_type_sections(input, &mut variables)?;

        // Parse SOS constraints section
        input = Self::parse_sos_section(input, &mut constraints, &mut variables)?;

        // Validate remaining input
        Self::validate_remaining_input(input)?;

        Ok(LpProblem { name, sense, objectives, constraints, variables })
    }
}

#[cfg(feature = "serde")]
impl<'de: 'a, 'a> serde::Deserialize<'de> for LpProblem<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(serde::Deserialize)]
        #[serde(field_identifier, rename_all = "lowercase")]
        enum Field {
            Constraints,
            Name,
            Objectives,
            Sense,
            Variables,
        }

        struct LpProblemVisitor<'a>(std::marker::PhantomData<LpProblem<'a>>);

        impl<'de: 'a, 'a> serde::de::Visitor<'de> for LpProblemVisitor<'a> {
            type Value = LpProblem<'a>;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("struct LpProblem")
            }

            fn visit_map<V: serde::de::MapAccess<'de>>(self, mut map: V) -> Result<LpProblem<'a>, V::Error> {
                let mut name: Option<Cow<'_, str>> = None;
                let mut sense = None;
                let mut objectives = None;
                let mut constraints = None;
                let mut variables = None;

                while let Some(key) = map.next_key()? {
                    match key {
                        Field::Name => {
                            if name.is_some() {
                                return Err(serde::de::Error::duplicate_field("name"));
                            }
                            name = map.next_value()?;
                        }
                        Field::Sense => {
                            if sense.is_some() {
                                return Err(serde::de::Error::duplicate_field("sense"));
                            }
                            sense = Some(map.next_value()?);
                        }
                        Field::Objectives => {
                            if objectives.is_some() {
                                return Err(serde::de::Error::duplicate_field("objectives"));
                            }
                            objectives = Some(map.next_value()?);
                        }
                        Field::Constraints => {
                            if constraints.is_some() {
                                return Err(serde::de::Error::duplicate_field("constraints"));
                            }
                            constraints = Some(map.next_value()?);
                        }
                        Field::Variables => {
                            if variables.is_some() {
                                return Err(serde::de::Error::duplicate_field("variables"));
                            }
                            variables = Some(map.next_value()?);
                        }
                    }
                }

                Ok(LpProblem {
                    name,
                    sense: sense.unwrap_or_default(),
                    objectives: objectives.unwrap_or_default(),
                    constraints: constraints.unwrap_or_default(),
                    variables: variables.unwrap_or_default(),
                })
            }
        }

        const FIELDS: &[&str] = &["name", "sense", "objectives", "constraints", "variables"];
        deserializer.deserialize_struct("LpProblem", FIELDS, LpProblemVisitor(std::marker::PhantomData))
    }
}

#[cfg(test)]
mod test {
    use std::borrow::Cow;

    use crate::model::{Coefficient, ComparisonOp, Constraint, Objective, Sense, Variable, VariableType};
    use crate::problem::LpProblem;

    const COMPLETE_INPUT: &str = "\\ This file has been generated by Author
\\ ENCODING=ISO-8859-1
\\Problem name: diet
Minimize
 obj1: -0.5 x - 2y - 8z
 obj2: y + x + z
 obj3: 10z - 2.5x
       + y
subject to:
c1:  3 x1 + x2 + 2 x3 = 30
c2:  2 x1 + x2 + 3 x3 + x4 >= 15
c3:  2 x2 + 3 x4 <= 25
bounds
x1 free
x2 >= 1
100 <= x2dfsdf <= -1
Integers
X31
X32
Generals
Route_A_1
Route_A_2
Route_A_3
Binary
V8
Semi-Continuous
 y
SOS
csos1: S1:: V1:1 V3:2 V5:3
csos2: S2:: V2:2 V4:1 V5:2.5
End";

    const SMALL_INPUT: &str = "\\ This file has been generated by Author
\\ ENCODING=ISO-8859-1
\\Problem name: diet
Minimize
 obj1: -0.5 x - 2y - 8z
 obj2: y + x + z
 obj3: 10z - 2.5x
       + y
subject to:
c1:  3 x1 + x2 + 2 x3 = 30
c2:  2 x1 + x2 + 3 x3 + x4 >= 15
c3:  2 x2 + 3 x4 <= 25
bounds
x1 free
x2 >= 1
100 <= x2dfsdf <= -1
End";

    #[test]
    fn test_small_input() {
        let problem = LpProblem::try_from(SMALL_INPUT).expect("test case not to fail");

        assert_eq!(problem.objectives.len(), 3);
        assert_eq!(problem.constraints.len(), 3);

        #[cfg(feature = "serde")]
        insta::assert_yaml_snapshot!(&problem, {
            ".objectives" => insta::sorted_redaction(),
            ".constraints" => insta::sorted_redaction(),
            ".variables" => insta::sorted_redaction()
        });
    }

    #[test]
    fn test_minified_example() {
        let problem = LpProblem::try_from(COMPLETE_INPUT).expect("test case not to fail");

        assert_eq!(problem.objectives.len(), 3);
        assert_eq!(problem.constraints.len(), 5);

        #[cfg(feature = "serde")]
        insta::assert_yaml_snapshot!(&problem, {
            ".objectives" => insta::sorted_redaction(),
            ".constraints" => insta::sorted_redaction(),
            ".variables" => insta::sorted_redaction()
        });
    }

    #[cfg(feature = "serde")]
    #[test]
    fn test_serialization_lifecycle() {
        let problem = LpProblem::try_from(COMPLETE_INPUT).expect("test case not to fail");
        let serialized_problem = serde_json::to_string(&problem).expect("test case not to fail");
        let _: LpProblem<'_> = serde_json::from_str(&serialized_problem).expect("test case not to fail");
    }

    #[test]
    fn test_add_variable() {
        let mut problem = LpProblem::new();
        let var = Variable::new("x1").with_var_type(VariableType::Binary);

        problem.add_variable(var);
        assert_eq!(problem.variable_count(), 1);
        assert!(problem.variables.contains_key("x1"));
    }

    #[test]
    fn test_add_constraint() {
        let mut problem = LpProblem::new();
        let constraint = Constraint::Standard {
            name: Cow::Borrowed("c1"),
            coefficients: vec![Coefficient { name: "x1", value: 1.0 }, Coefficient { name: "x2", value: 2.0 }],
            operator: ComparisonOp::LTE,
            rhs: 5.0,
        };

        problem.add_constraint(constraint);
        assert_eq!(problem.constraint_count(), 1);
        assert_eq!(problem.variable_count(), 2);
    }

    #[test]
    fn test_add_objective() {
        let mut problem = LpProblem::new().with_sense(Sense::Minimize).with_problem_name(Cow::Borrowed("test"));
        let objective = Objective {
            name: Cow::Borrowed("obj1"),
            coefficients: vec![Coefficient { name: "x1", value: 1.0 }, Coefficient { name: "x2", value: -1.0 }],
        };

        problem.add_objective(objective);
        assert_eq!(problem.objective_count(), 1);
        assert_eq!(problem.variable_count(), 2);
    }
}
