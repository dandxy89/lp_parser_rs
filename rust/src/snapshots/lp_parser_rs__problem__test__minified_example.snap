---
source: rust/src/problem.rs
expression: "&problem"
---
name: "Problem name: diet"
sense: Minimize
objectives:
  obj1:
    name: obj1
    coefficients:
      - name: x
        value: -0.5
      - name: y
        value: -2
      - name: z
        value: -8
  obj2:
    name: obj2
    coefficients:
      - name: y
        value: 1
      - name: x
        value: 1
      - name: z
        value: 1
  obj3:
    name: obj3
    coefficients:
      - name: z
        value: 10
      - name: x
        value: -2.5
      - name: y
        value: 1
constraints:
  c1:
    type: Standard
    name: c1
    coefficients:
      - name: x1
        value: 3
      - name: x2
        value: 1
      - name: x3
        value: 2
    operator: EQ
    rhs: 30
  c2:
    type: Standard
    name: c2
    coefficients:
      - name: x1
        value: 2
      - name: x2
        value: 1
      - name: x3
        value: 3
      - name: x4
        value: 1
    operator: GTE
    rhs: 15
  c3:
    type: Standard
    name: c3
    coefficients:
      - name: x2
        value: 2
      - name: x4
        value: 3
    operator: LTE
    rhs: 25
  csos1:
    type: SOS
    name: csos1
    sos_type: S1
    weights:
      - name: V1
        value: 1
      - name: V3
        value: 2
      - name: V5
        value: 3
  csos2:
    type: SOS
    name: csos2
    sos_type: S2
    weights:
      - name: V2
        value: 2
      - name: V4
        value: 1
      - name: V5
        value: 2.5
variables:
  Route_A_1:
    name: Route_A_1
    var_type: General
  Route_A_2:
    name: Route_A_2
    var_type: General
  Route_A_3:
    name: Route_A_3
    var_type: General
  V1:
    name: V1
    var_type: SOS
  V2:
    name: V2
    var_type: SOS
  V3:
    name: V3
    var_type: SOS
  V4:
    name: V4
    var_type: SOS
  V5:
    name: V5
    var_type: SOS
  V8:
    name: V8
    var_type: Binary
  X31:
    name: X31
    var_type: Integer
  X32:
    name: X32
    var_type: Integer
  x:
    name: x
    var_type: Free
  x1:
    name: x1
    var_type: Free
  x2:
    name: x2
    var_type:
      LowerBound: 1
  x2dfsdf:
    name: x2dfsdf
    var_type:
      DoubleBound:
        - 100
        - -1
  x3:
    name: x3
    var_type: Free
  x4:
    name: x4
    var_type: Free
  y:
    name: y
    var_type: SemiContinuous
  z:
    name: z
    var_type: Free
